### PR TITLE
indexeddb: Shrink values in inbound_group_sessions store, improving performance all around

### LINKED ISSUE #3057: Indexed DB performance is slow because of large values
We've noticed Element Web R is a little slow at some operations (e.g. [resetting key backup](https://github.com/element-hq/element-web/issues/26892), [exporting keys](https://github.com/matrix-org/matrix-rust-sdk/issues/2914), [importing keys](https://github.com/vector-im/element-web/issues/26692), migrating Element R schemata, upgrading from legacy to rust crypo).

We tracked the worst slowdowns to slow operations on the `inbound_group_sessions2` store in Indexed DB.

I did some research and discovered that there are some rules of thumb we can use to have faster performance in Indexed DB: [keep your Indexed DB keys and values small if you want good performance!](https://www.artificialworlds.net/blog/2024/01/24/keep-your-indexed-db-keys-and-values-small-if-you-want-good-performance/) and [don’t store arrays of numbers in Indexed DB – use base64 instead](https://www.artificialworlds.net/blog/2024/01/25/dont-store-arrays-of-numbers-in-indexed-db-use-base64-instead/).

So, we should apply these rules to our Indexed DB stores, starting with `inbound_group_sessions2`.

Currently, serializing an `InboundGroupSession` looks like:

* `matrix_sdk_store_encryption::StoreCipher::encrypt_value_typed` -> `EncryptedValue`
* `matrix_sdk_store_encryption::StoreCipher::encrypt_value` -> `Vec<u8>` (serialized EncryptedValue)
* `matrix_sdk_indexeddb::crypto_store::indexeddb_serializer::IndexeddbSerializer::serialize_value_as_bytes` -> `Vec<u8> (serialized EncryptedValue)
* `matrix_sdk_indexeddb::crypto_store::IndexeddbCryptoStore::serialize_inbound_group_session` -> `JsValue` (created from an InboundGroupSessionIndexedDbObject that has the `Vec<u8>` (serialized EncryptedValue) as a property)

There are two ways we can improve this:

1. Avoid double-serializing: currently the session is a serialized `InboundGroupSessionIndexedDbObject` that contains a serialized `EncryptedValue`. This makes the resulting object stored in IndexedDb much larger than it needs to me.
2. Avoid use of arrays of ints in the values we store. Instead, encode them as base64 strings.

So I propose to change serializing `InboundGroupSession` to something like:

* `matrix_sdk_store_encryption::StoreCipher::encrypt_value_base64_typed` -> `EncryptedValueBase64` (same as `EncryptedValue` except it contains `String`s instead of `Vec<u8>` and `[u8; N]`).
* `matrix_sdk_indexeddb::crypto_store::indexeddb_serializer::IndexeddbSerializer::maybe_encrypt_value` -> `MaybeEncrypted` (not-serialized wrapper around `EncryptedValue` representing the fact that in some cases we don't encrypt.)
* `matrix_sdk_indexeddb::crypto_store::IndexeddbCryptoStore::serialize_inbound_group_session` -> `JsValue` (created from an InboundGroupSessionIndexedDbObject that has the `MaybeEncrypted` as a property)

This will make the values we store much smaller because it avoids double-encoding, and it uses base64 instead of arrays of ints. Both of these should provide good performance improvements.

Of course, we will need a migration from the old schema to the new one.

In addition, since we are changing the schema, I plan to include other changes that make us compatible with future planned changes for [Resetting key backup #26892](https://github.com/element-hq/element-web/issues/26892).

### PR DESCRIPTION
**Pull request by andybalaam**
Fixes #3057.

Change the way we store values in Indexed DB for the crypto store, specifically in the `inbound_group_sessions` store. Reduce the size of the stored values and replace arrays of numbers with base 64 strings. The rationale and some implementation ideas are documented in #3057. (TL;DR: small values are faster, base64 is faster.)

This change:

* Introduces a `MaybeEncrypted` type in `matrix_sdk_indexeddb::crypto_store::indexeddb_serializer` that encapsulates the fact that we serialize differently if the session is encrypted with a store cipher. (Previously, we just serialized an array of numbers in both cases. This was actually quite confusing since these arrays of numbers were holding encoded versions of very different values.) This enum is `untagged` in Serde to avoid increasing the size of Indexed DB records.
* Adds `matrix_sdk_store_encryption::EncryptedValueBase64`, which works the same way as `EncryptedValue`, but stores arrays of numbers as base 64.
* Adds a migration to the new schema to `matrix_sdk_indexeddb::crypto_store::migrations`, including adding the `backed_up_to` property, so we can support a fix for #26892 (see this [comment](https://github.com/element-hq/element-web/issues/26892#issuecomment-1906336076)) without doing another migration.

I also wrote a performance test that runs in a web browser. In the checked-in version, it runs for 2K records, but when I manually increased the number on my machine to 200K, I got these results:

```
Inserting 200000 records with v8 schema took 216,745ms.
Inserting 200000 records with v10 schema took 34,875ms.

Counting 200000 records with v8 schema took 11,786ms.
Counting 200000 records with v10 schema took 469ms.
```

So the speedup is impressive.

### CODE REVIEW

**Review by BillCarsonFr — APPROVED** [review]
Great work, thx!
LGTM to me. I agree that it's worth it to include the future `backed_up_to` field at this point.

File: crates/matrix-sdk-store-encryption/src/lib.rs [hunk 1]
```diff
@@ -674,7 +674,43 @@ impl StoreCipher {
...
+    /// # Arguments
+    ///
+    /// * `value` - The EncryptedValueBase64 of a value that should be
+    ///   decrypted.
+    ///
+    /// The method will return the raw decrypted value
+    ///
+    /// # Examples
+    ///
+    /// ```
```
Comment (Hywan) [inline crates/matrix-sdk-store-encryption/src/lib.rs:692]: You can use `rust,no_run` if you still want to assume it's Rust code, but you don't want to run it (only check it).

**Review by Hywan — APPROVED** [review]
It looks excellent to me. I've a couple of really tiny feedback, but overall, great work!

File: crates/matrix-sdk-indexeddb/src/crypto_store/indexeddb_serializer.rs [hunk 2]
```diff
@@ -161,4 +189,23 @@ impl IndexeddbSerializer {
             serde_json::from_slice(value).map_err(CryptoStoreError::backend)
         }
     }
+
+    /// Decode a value that was previously encoded with
+    /// [`maybe_encrypt_value`]
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-indexeddb/src/crypto_store/indexeddb_serializer.rs:194]: This change allows my editor to jump to definition, when it failed before, so I reckon it's good. I've changed it for all items in this file. Thanks! 9e5bcbcfd7

File: crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs [hunk 3]
```diff
@@ -18,16 +18,19 @@ use tracing::{debug, info};
...
 };
 
 mod old_keys {
     /// Old format of the inbound_group_sessions store which lacked indexes or a
     /// sensible structure
     pub const INBOUND_GROUP_SESSIONS_V1: &str = "inbound_group_sessions";
+
+    /// inbound_group_sessions2 with large values in each record due to double
+    /// JSON-encoding and arrays of ints instead of base64.
+    /// Also lacked the backed_up_to property+index.
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs:32]: Thanks, done in 47c20f708c

File: crates/matrix-sdk-store-encryption/src/lib.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk-store-encryption/src/lib.rs:692]: Thanks, added to all in the file. 885d85740c

File: crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs [hunk 4]
```diff
@@ -18,16 +18,19 @@ use tracing::{debug, info};
...
+    crypto_store::{indexeddb_serializer::IndexeddbSerializer, keys, Result},
     IndexeddbCryptoStoreError,
 };
 
 mod old_keys {
     /// Old format of the inbound_group_sessions store which lacked indexes or a
     /// sensible structure
     pub const INBOUND_GROUP_SESSIONS_V1: &str = "inbound_group_sessions";
+
+    /// inbound_group_sessions2 with large values in each record due to double
```
[the remark this answers is not in the document]
Comment (andybalaam) [reply crates/matrix-sdk-indexeddb/src/crypto_store/migrations.rs:30]: Thanks, done in 64fb43602c

**Comment by poljar** [comment]
Oh no: https://github.com/matrix-org/matrix-rust-sdk/issues/718.

**Comment by dkasak** [comment]
Great work with the speed-up!

I'm a bit late to the party, but what role does base64-encoding the payload serve in this? It seems that the trick is to replace an array with lots of values with a single value (a string), which could've just as well been a JSON string. Which would mean the base64 step only serves to add overhead on top of this?

The only way the base64 encoding would make sense is if we opted for a more compact binary representation of the array of numbers (which I think was the ultimate goal of #718 there which poljar linked).

**Comment by andybalaam** [comment]
> Oh no: #718.

@poljar hmm. I think we could move the JSON-specific stuff into the indexeddb crate if we wanted to?

**Comment by andybalaam** [comment]
> Great work with the speed-up!
> 
> I'm a bit late to the party, but what role does base64-encoding the payload serve in this? It seems that the trick is to replace an array with lots of values with a single value (a string), which could've just as well been a JSON string. Which would mean the base64 step only serves to add overhead on top of this?
> 
> The only way the base64 encoding would make sense is if we opted for a more compact binary representation of the array of numbers (which I think was the ultimate goal of #718 there which poljar linked).

@dkasak the base64 encoding is to transform an array of bytes into a string, because [I found](https://www.artificialworlds.net/blog/2024/01/25/dont-store-arrays-of-numbers-in-indexed-db-use-base64-instead/) that strings are much faster than arrays of numbers to store inside Indexed DB. When you say a JSON string, do you mean treat the bytes as characters directly? That would result in invalid UTF-8, which is not allowed in JSON, right?

**Comment by poljar** [comment]
> > Oh no: #718.
> 
> @poljar hmm. I think we could move the JSON-specific stuff into the indexeddb crate if we wanted to?

Yeah, maybe, will think about it.

**Comment by andybalaam** [comment]
We are already encoding to JSON using `to_string` the question is _what_ are we encoding? Previously it was a `Vec<u8>` and now it's a `String`, which contains base64.

**Comment by andybalaam** [comment]
If I remember correctly, the `Vec<u8>` we have is a pickled version of an `InboundGroupSession`.

**Comment by dkasak** [comment]
Just to write down the conclusion: after a quick chat with @andybalaam, only the [unencrypted case](https://github.com/matrix-org/matrix-rust-sdk/pull/3073/files#diff-514d0599fe785ee0062f5a4df3f42b32a0c8ace004a8368e62b08321e18b1e56R158) unnecessarily base64-encodes the payload before storing. The [encrypted case](https://github.com/matrix-org/matrix-rust-sdk/pull/3073/files#diff-514d0599fe785ee0062f5a4df3f42b32a0c8ace004a8368e62b08321e18b1e56R155) a couple of lines above does what is expected, so that's why we were talking past each other.

**Pull request merged**